sky/cloud_stores.py

```diff
@@ -215,8 +215,74 @@ def get_storage_from_path(url: str) -> CloudStorage:
     return _REGISTRY[result.scheme]
 
 
+class IBMCosCloudStorage(CloudStorage):
+    """IBM Cloud Storage."""
+    # install rclone if package isn't already installed
+    # pylint: disable=line-too-long
+    _GET_RCLONE = [
+        'rclone version >/dev/null 2>&1 || curl https://rclone.org/install.sh | sudo bash',
+    ]
+
+    def is_directory(self, url: str) -> bool:
+        """Returns whether IBM cos bucket's 'url' is a directory.
+
+        In cloud object stores, a "directory" refers to a regular object whose
+        name is a prefix of other objects.
+        """
+
+        region = url.split('//')[1].split('/')[0]
+        s3 = ibm.get_cos_resource(region)
+        bucket_name, path, _ = data_utils.split_cos_path(url)
+        bucket = s3.Bucket(bucket_name)
+
+        num_objects = 0
+        for obj in bucket.objects.filter(Prefix=path):
+            print(obj)
+
+            num_objects += 1
+            if obj.key == path:
+                return False
+            # If there are more than 1 object in filter, then it is a directory
+            if num_objects == 3:
+                return True
+
+        # A directory with few or no items
+        return True
+
+    def get_cmd(self, source: str, destination: str):
+        bucket_name, data_path, bucket_region = data_utils.split_cos_path(
+            source)
+        data_path_in_bucket = bucket_name + data_path
+        rclone_config_data = data_utils.store_rclone_config(
+            bucket_name, bucket_region, False)
+        # configure_rclone cmd stores bucket profile in rclone config file at the cluster's nodes.
+        configure_rclone = (
+            f' mkdir -p ~/.config/rclone/ && echo "{rclone_config_data}">> {data_utils.RCLONE_CONFIG_PATH}'
+        )
+        # bucket_name also serves as a profile name in rclone.conf
+        download_via_rclone = (
+            f'rclone copy {bucket_name}:{data_path_in_bucket} {destination}')
+
+        all_commands = list(self._GET_RCLONE)
+        all_commands.append(configure_rclone)
+        all_commands.append(download_via_rclone)
+        return ' && '.join(all_commands)
+
+    def make_sync_dir_command(self, source: str, destination: str) -> str:
+        """Downloads a directory from 'source' bucket to remote vm
+          at 'destination' using rclone."""
+        return self.get_cmd(source, destination)
+
+    def make_sync_file_command(self, source: str, destination: str) -> str:
+        """Downloads a file from 'source' bucket to remote vm
+          at 'destination' using rclone."""
+        return self.make_sync_dir_command(source, destination)
```

Comment:
`make_sync_dir_command` is an overwritten interface method (`CloudStorage`). Can't change its name.

Reply:
Sorry for lack of explanation. I didn't meant to say to change the name of `make_sync_dir_command`. If you see, `make_sync_file_command` just simply returns `make_sync_dir_command`, and `make_sync_dir_command` just returns `get_cmd` (which is now `_get_rclone_sync_command`). So, I was suggesting to return `_get_rclone_sync_command` rather than returning `make_sync_dir_command` from `make_sync_file_command` :)

Reply:
Understood, but making `make_sync_file_command` invoke `make_sync_dir_command` was deliberate. It:
- Ensures future contributors will notice these two methods are running the same code.
- Removes duplicate code. Currently it's mild, but the implementation can easily grow.
Can you please describe the reasoning behind the change?

Reply:
Also in a view of future contributor, to see what `make_sync_file_command` does, they would take a look at `make_sync_dir_command`, and then move on to `_get_rclone_sync_command` when they can move on to `_get_rclone_sync_command` directly. And this change should also allow them to notice two methods running the same code. But your point makes sense as well so I'll leave this to your decision :))